frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/toolbar/DiscoverAction.java

```diff
@@ -81,40 +84,52 @@ public void actionPerformed(AnActionEvent anActionEvent) {
   }
 
+  @UIEffect
   private static void openMacheteFile(Project project, GitRepository gitRepository) {
-    GuiUtils.invokeLaterIfNeeded(() -> {
-      var file = GitVfsUtils.getMacheteFile(gitRepository);
-      if (file.isDefined()) {
-        OpenFileAction.openFile(file.get(), project);
-      } else {
-        VcsNotifier.getInstance(project)
-            .notifyError(
-                /* title */ getString("action.GitMachete.OpenMacheteFileAction.notification.title.machete-file-not-found"),
-                /* message */ format(
-                    getString("action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found"),
-                    gitRepository.getRoot().getPath()));
-      }
-    }, NON_MODAL);
+    var file = GitVfsUtils.getMacheteFile(gitRepository);
+    if (file.isDefined()) {
+      OpenFileAction.openFile(file.get(), project);
+    } else {
+      VcsNotifier.getInstance(project)
+          .notifyError(
+              /* title */ getString("action.GitMachete.OpenMacheteFileAction.notification.title.machete-file-not-found"),
+              /* message */ format(
+                  getString("action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found"),
+                  gitRepository.getRoot().getPath()));
+    }
   }
 
   private void saveDiscoveredLayout(IGitMacheteRepositorySnapshot repositorySnapshot, Path macheteFilePath, Project project,
-      BaseEnhancedGraphTable baseEnhancedGraphTable, IBranchLayoutWriter branchLayoutWriter) {
+      BaseEnhancedGraphTable baseEnhancedGraphTable, IBranchLayoutWriter branchLayoutWriter, @UI Runnable postWriteRunnable) {
     var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
     if (branchLayout == null) {
       VcsNotifier.getInstance(project).notifyError(
           /* title */ getString("action.GitMachete.DiscoverAction.notification.title.cannot-discover-layout-error"),
           /* message */ "");
       return;
     }
+
     new Task.Backgroundable(project, getString("action.GitMachete.DiscoverAction.write-file.task-title")) {
       @Override
       public void run(ProgressIndicator indicator) {
-        Try.run(() -> branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupOldLayout */ true))
-            .onFailure(e -> VcsNotifier.getInstance(project).notifyError(
-                /* title */ getString("action.GitMachete.DiscoverAction.notification.title.write-file-error"),
-                /* message */ e.getMessage() != null ? e.getMessage() : ""))
-            .onSuccess(__ -> baseEnhancedGraphTable.queueRepositoryUpdateAndModelRefresh());
+        Try.run(() -> branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupOldLayout */ true));
+      }
+
+      @Override
+      public void onSuccess() {
+        baseEnhancedGraphTable.queueRepositoryUpdateAndModelRefresh();
+        VfsUtil.markDirtyAndRefresh(/* async */ false, /* recursive */ true, /* reloadChildren */ false,
```

Comment:
> Nit: hmm can this be moved into `Task`'s own `onFailure`/`onSuccess`? Then probably `GuiUtils.invokeLaterIfNeeded` in `openMacheteFile` won't be even needed since IIRC both these `Task`'s callbacks are executed on UI thread; `openMacheteFile` can be `@UIEffect` and `postWriteRunnable` can be `@UI`
~ @PawelLipski 

I tried it before but splitting the code into `onThrowable` and `onSuccess` has not resolved the issue.
However, I managed to do that now. Actually, for some reason, a refresh of VFS before opening the machete file is required when `write` and `postWriteRunnable` (open) are split between the methods.

Works fine to me but I would appreciate it if someone else could verify it too.

Reply:
Whoops actually... does this `markDirtyAndRefresh` thing also **reparse** the files, or only syncs them with the disk? Coz I don't want the issues like https://github.com/VirtusLab/git-machete-intellij-plugin/issues/679 to recur 😱

Reply:
checked, `markDirtyAndRefresh` does no call `FileContentUtilCore#reparseFiles`